src/services/organizationService.ts

```diff
@@ -0,0 +1,209 @@
+    O.map(({ page, pageSize }) => ({
+      offset: page * pageSize,
+      limit: pageSize,
+    })),
+    O.getOrElse(() => ({}))
+  );
+
+export const getOrganizations = (
+  page?: number,
+  pageSize?: number,
+  searchQuery?: string
+): TE.TaskEither<Error, Organizations> =>
+  pipe(
+    TE.tryCatch(
+      () =>
+        OrganizationModel.findAll({
+          include: [OrganizationModel.associations.referents],
+          ...filterByNameOrFiscalCode(searchQuery),
+          ...paging(page, pageSize),
+        }),
+      E.toError
+    ),
+    TE.map(
+      flow(
+        AR.map((m) =>
+          OrganizationWithReferents.decode({
+            keyOrganizationFiscalCode: m.fiscalCode,
+            organizationFiscalCode: m.fiscalCode,
+            organizationName: m.name,
+            pec: m.pec,
+            insertedAt: m.insertedAt,
+            referents: m.referents.map((r) => r.fiscalCode),
+          })
+        ),
+        AR.filter(E.isRight),
+        AR.map((e) => e.right)
+      )
+    )
+  );
+
+export const upsertOrganization = (
+  organizationWithReferents: OrganizationWithReferents
+): TE.TaskEither<Error, OrganizationWithReferents> =>
+  pipe(
+    TE.of(
+      organizationWithReferents.keyOrganizationFiscalCode !==
+        organizationWithReferents.organizationFiscalCode
+    ),
+    TE.chain((shouldUpdatePrimaryKey) =>
+      TE.tryCatch(async () => {
+        shouldUpdatePrimaryKey
+          ? await OrganizationModel.sequelize.query(
+              UpdateOrganizationPrimaryKey,
+              {
+                raw: true,
+                replacements: {
+                  new_fiscal_code:
+                    organizationWithReferents.organizationFiscalCode,
+                  old_fiscal_code:
+                    organizationWithReferents.keyOrganizationFiscalCode,
+                },
+                type: QueryTypes.UPDATE,
+              }
+            )
+          : void 0;
+      }, E.toError)
+    ),
+    TE.chain(() =>
+      TE.tryCatch(
+        async () =>
+          await OrganizationModel.upsert({
+            fiscalCode: organizationWithReferents.organizationFiscalCode,
+            name: organizationWithReferents.organizationName,
+            pec: organizationWithReferents.pec,
+          }),
+        E.toError
+      )
+    ),
+    TE.chain(([organization, _]) =>
+      pipe(
+        TE.tryCatch(async () => {
+          const referentsToRemove = await organization.getReferents();
+          await organization.removeReferents(referentsToRemove, {
+            force: true,
+          });
+          const referents = (
+            await Promise.all(
+              organizationWithReferents.referents.map((r) =>
+                Referent.upsert({ fiscalCode: r })
+              )
+            )
+          ).map(([r, _]) => r);
+          await organization.addReferents(referents, {
+            ignoreDuplicates: true,
+          });
+        }, E.toError),
```

Comment:
why not refactor this implementation into a chain based one?

Reply:
in https://github.com/pagopa/ade-aa-ms-mock/pull/16/commits/9bd4e8454fe620851cd808de302de858df4fb62c